spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/MessageConverterConfigurer.java

```diff
@@ -165,50 +178,65 @@ private boolean isNativeEncodingNotSet(ProducerProperties producerProperties, Co
 		}
 	}
 
-	private PartitionKeyExtractorStrategy getPartitionKeyExtractorStrategy(
-			ProducerProperties producerProperties) {
+	@SuppressWarnings("deprecation")
+	private PartitionKeyExtractorStrategy getPartitionKeyExtractorStrategy(ProducerProperties producerProperties) {
+		PartitionKeyExtractorStrategy partitionKeyExtractor;
 		if (producerProperties.getPartitionKeyExtractorClass() != null) {
-			return getBean(producerProperties.getPartitionKeyExtractorClass().getName(),
-					PartitionKeyExtractorStrategy.class);
+			logger.warn("'partitionKeyExtractorClass' option is deprecated as of v2.0. Please configure partition "
+					+ "key extractor as a @Bean that implements 'PartitionKeyExtractorStrategy'. Additionally you can "
+					+ "specify 'spring.cloud.stream.bindings.output.producer.partitionKeyExtractorName' to specify which "
+					+ "bean to use in the event there are more then one.");
+			partitionKeyExtractor = instantiate(producerProperties.getPartitionKeyExtractorClass(), PartitionKeyExtractorStrategy.class);
+		}
+		else if (StringUtils.hasText(producerProperties.getPartitionKeyExtractorName())) {
+			partitionKeyExtractor = this.partitionKeyExtractors.get(producerProperties.getPartitionKeyExtractorName());
+			Assert.notNull(partitionKeyExtractor, "PartitionKeyExtractorStrategy bean with the name '" + producerProperties.getPartitionKeyExtractorName()
+				+ "' can not be found. Has it been configured (e.g., @Bean)?");
 		}
-		return null;
+		else {
+			Assert.isTrue(CollectionUtils.isEmpty(this.partitionKeyExtractors) || this.partitionKeyExtractors.size() <= 1,
```

Comment:
I think you misunderstood - we don't need the isEmpty check, it's handled by <= 1 - I can fix on merge if you agree.

Reply:
oops, yeah, please! Thank you